languages/elixir/exercises/concept/pattern-matching/.meta/example.ex

```diff
@@ -0,0 +1,19 @@
+defmodule DateString do
+
+  def from_tuple({{year, month, day}, {hour, minute, second}}) do
+    "#{year}-#{month}-#{day} #{hour}:#{minute}:#{second}"
+  end
+
+  def from_map(
+        %{
+          year: year,
+          month: month,
+          day: day,
+          hour: hour,
+          minute: minute,
+          second: second,
+        }
+      ) do
+    "#{year}-#{month}-#{day} #{hour}:#{minute}:#{second}"
+  end
+end
```

Comment:
Is there anything else we can add to this to increase complexity at all?

I think these are great for teaching the destructuring use of pattern matching, but is there something else we can add to this? input validity checks? guards?

Reply:
I was thinking of refactoring the F# exercise for introducing guards.  For the pattern-matching concept here I was trying to come keep things modular so that they might conceivably be done in isolation, but as we've discussed, isolating something as fundamental as pattern matching is difficult.  

I'm kinda seeing several of these exercises being a critical "pre-requisite" chunk for the rest of the track.  That would include the intro to pattern-matching, the head|tail syntax, multi-clause functions, and guard clauses.  If we wanted to "bundle" a few exercises together, I'm not sure how we should notate that ... `pattern-match-1`, `pattern-match-2`?  Just thinking out loud.

Reply:
Ya, really after some basic types, this exercise is the real gateway to the rest of the track.

Head-tail syntax discussion in `lists` is limited to appending to the head of a list, so it very well can be expended on in a pattern matching discussion.

The name of the concept slug doesn't so much matter as long as it is unique kebab case.  the progression is determined in the track's config.json where an exercise lists its direct pre-requisite to allow access to it.

Reply:
> I'm kinda seeing several of these exercises being a critical "pre-requisite" chunk for the rest of the track. That would include the intro to pattern-matching, the head|tail syntax, multi-clause functions, and guard clauses. If we wanted to "bundle" a few exercises together, I'm not sure how we should notate that ... pattern-match-1, pattern-match-2? Just thinking out loud.

> Ya, really after some basic types, this exercise is the real gateway to the rest of the track.

I would suggest this exercise to introduce the concept of pattern matching: what is it, when to use it, how to use it, etc. For the [F# pattern matching exercise](https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/pattern-matching/.meta/design.md), I could rather easily do that, as pattern matching there also worked on constant values. I also introduced wildcard pattern matching, guard clauses and exhaustiveness checking. I'm not sure which of these concepts also apply to Elixir, but it is worth considering to introduce them here. 

The reason why I wanted to not explain how to do pattern matching on lists or tuples in the F# pattern matching exercise, is that I felt it made more sense to introduce them in the [lists exercise](https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/lists/.meta/design.md) respectively the tuples exercise. Would something like this be possible for Elixir too?

Another option would be to have this pattern matching exercise just depend on the lists concept, and then introduce pattern matching on lists here too (besides basic pattern matching). Combining that with guards, wildcards and such could make this a full fleshed exercise too. You could then have tuple pattern matching introduced in the tuples exercise.

Note that I struggled with these pattern matching dependencies a bit too, but I'm quite happy with how it turned out for the F# track.